magpie-core/src/main/java/io/openraven/magpie/core/cspm/services/PolicyAnalyzerServiceImpl.java

```diff
@@ -54,74 +52,104 @@ public ScanResults analyze(List<PolicyContext> policies) throws Exception {
   }
 
+  private void processPolicy(Policy policy,
+                             AtomicInteger numOfViolations,
+                             List<Violation> policyViolations,
+                             Map<Rule, ScanResults.IgnoredReason> policyIgnoredRules) {
+    if (!policy.isEnabled()) { // Not enabled
+      LOGGER.info("Policy '{}' disabled", policy.getPolicyName());
+      return;
+    }
+
+    if (!cloudProviderAssetsAvailable(policy)) { // No cloud related assets
+      LOGGER.warn("There was no assets detected for cloudProvider: {}, policy: {}",
+        policy.getCloudProvider(), policy.getPolicyName());
+      return;
+    }
+
+    LOGGER.info("Analyzing policy - {}", policy.getPolicyName());
+    policy.getRules().forEach(rule -> {
+      executeRule(numOfViolations, policyViolations, policyIgnoredRules, policy, rule);
+    });
+  }
+
+  private boolean cloudProviderAssetsAvailable(Policy policy) {
+    return jdbi.withHandle(handle -> handle.createQuery(String.format(EXIST_ASSETS_PER_CLOUD, policy.getCloudProvider()))
+      .mapTo(Boolean.class).findFirst().orElseThrow());
+  }
+
+  private void executeRule(AtomicInteger numOfViolations,
+                           List<Violation> policyViolations,
+                           Map<Rule, ScanResults.IgnoredReason> policyIgnoredRules,
+                           Policy policy,
+                           Rule rule) {
+    if (!rule.isEnabled()) {
+      policyIgnoredRules.put(rule, ScanResults.IgnoredReason.DISABLED);
+      LOGGER.info("Rule '{}' disabled", rule.getRuleName());
+      return;
+    }
+
+    if (rule.isManualControl()) {
+      policyIgnoredRules.put(rule, ScanResults.IgnoredReason.MANUAL_CONTROL);
+      LOGGER.info("Rule not analyzed (manually controlled) - {}", rule.getRuleName());
+      return;
+    }
+
+    var missingAssets = checkForMissingAssets(jdbi, rule.getSql());
+    if (!missingAssets.isEmpty()) { // Missing assets found
+      policyIgnoredRules.put(rule, ScanResults.IgnoredReason.MISSING_ASSET);
+      LOGGER.info("Missing assets for analyzing the rule, ignoring. [assets={}, rule={}]", missingAssets, rule.getRuleName());
+      return;
+    }
+
+    LOGGER.info("Analyzing rule - {}", rule.getRuleName());
+    LocalDateTime evaluatedAt = LocalDateTime.now();
+
+    var results = jdbi.withHandle(handle -> handle.createQuery(rule.getSql()).mapToMap().list());
+
+    StringWriter evalOut = new StringWriter();
+    StringWriter evalErr = new StringWriter();
+    if (!Optional.ofNullable(rule.getEval()).orElse("").isEmpty()) {
+      try {
+        evalOut.append(evaluate(rule, results));
+      } catch (Exception e) {
+        evalErr.append(e.getMessage());
+      }
+    }
+
+    results.forEach(result -> {
+      Violation violation = new Violation();
+      violation.setPolicyId(policy.getId());
+      violation.setRuleId(rule.getId());
+      violation.setAssetId(result.get("asset_id").toString());
```

Comment:
@kickroot I have doubts about line below: we are producing violation information based on policy.description and not rule.description which was violated,
which will then lead to similar violation info for various rules

Reply:
But in the end I see that ReportServiceImpl never getting this data for building reports

Reply:
Don't worry about this for the moment, I'll verify the output columns separately.